### PR TITLE
Add support for Optical Duplicate tracking to MarkDuplicatesDataflow

### PR DESCRIPTION
**Pull request by tovanadler**
Reverts broadinstitute/hellbender#748

### CODE REVIEW

**Comment by tovanadler** [comment]
Sorry about all the back and forth! This should be ready for review. @akiezun

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 1]
```diff
@@ -71,10 +100,19 @@ protected void setupPipeline(final Pipeline pipeline) {
         final PCollectionView<SAMFileHeader> headerPcolView = pipeline.apply(Create.of(header)).apply(View.<SAMFileHeader>asSingleton());
 
         final PCollection<GATKRead> preads = readsSource.getReadPCollection(intervals);
-        final PCollection<GATKRead> results = preads.apply(new MarkDuplicatesDataflowTransform(headerPcolView));
+
+        final OpticalDuplicateFinder finder = READ_NAME_REGEX != null ?
+            new OpticalDuplicateFinder(OpticalDuplicateFinder.DEFAULT_READ_NAME_REGEX, OPTICAL_DUPLICATE_PIXEL_DISTANCE, null) : null;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Shouldn't this be `READ_NAME_REGEX` instead of `DEFAULT_READ_NAME_REGEX`?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 2]
```diff
@@ -124,37 +170,42 @@ public int partitionFor(final GATKRead read, final int n) {
...
-         * (3) Paired ends are grouped by a similar key as above but using both reads.
-         * (4) Any unpaired end is emitted, unmodified.
-         * (5) The remained paired ends are scored and all but the highest scoring are marked as
-         *     duplicates. Both reads in the pair are emitted.
+         * (1) labelReadsByName: label each read with its read group and read name.
+         * (2) GroupByKey: group together reads with the same group and name.
+         * (3) labelPairedEndsWithAlignmentInfo:
+         *   (a) Sort each group of reads (see LexicographicSort below).
+         *   (b) Pair consecutive reads into PairedEnds. In most cases there will only be two reads
+         *       with the same name. TODO: explain why there might be more.
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Could you open a ticket to add this explanation?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 3]
```diff
@@ -124,37 +170,42 @@ public int partitionFor(final GATKRead read, final int n) {
...
-            final PTransform<PCollection<? extends KV<String, Iterable<GATKRead>>>, PCollection<KV<String, PairedEnds>>> markGroupedDuplicatePairs = markGroupedDuplicatePairs(headerPcolView);
-            final PTransform<PCollection<? extends KV<String, Iterable<PairedEnds>>>, PCollection<GATKRead>> markPairedEnds = markPairedEnds();
+            final PTransform<PCollection<? extends GATKRead>, PCollection<KV<String, GATKRead>>> labelReadsByName = labelReadsByName(headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<GATKRead>>>, PCollection<KV<String, PairedEnds>>> labelPairedEndsWithAlignmentInfo =
+                labelPairedEndsWithAlignmentInfo(headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<PairedEnds>>>, PCollection<GATKRead>> markDuplicatePairedEnds = markDuplicatePairedEnds();
 
             return pairs
-                    .apply(makeKeysForPairs)
+                    .apply(labelReadsByName)
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: I think including key in the name is useful to help mentally map back to what's going on in dataflow.   Maybe `keyReadsByName`?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 4]
```diff
@@ -419,11 +733,12 @@ public PairedEnds decode( InputStream inStream, Context context ) throws CoderEx
     }
 
     // TODO: extract this into an independent class, and unify with other comparators in the codebase
-    private static final class CoordinateOrder implements Comparator<GATKRead>, Serializable {
+    // TODO: explain how this ordering works.
+    private static final class LexicographicOrder implements Comparator<GATKRead>, Serializable {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Heads up, this got renamed to GATKOrder

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 5]
```diff
@@ -218,20 +269,48 @@ public void processElement(final ProcessContext context) throws Exception {
                             }
 
                             //order by score
-                            final List<PairedEnds> scored = Ordering.natural().reverse().onResultOf((PairedEnds pair) -> pair.score()).immutableSortedCopy(paired.get(true));
+                            List<PairedEnds> scored = Ordering.natural().reverse().onResultOf((PairedEnds pair) -> pair.score()).sortedCopy(paired.get(true));
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: prefer  a method reference to a lambda when possible.  `onResultOf((PairedEnds pair) -> pair.score())` -> `onResultOf(PairedEnds::score)`

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 6]
```diff
@@ -218,20 +269,48 @@ public void processElement(final ProcessContext context) throws Exception {
...
                             final PairedEnds best = Iterables.getFirst(scored, null);
-                            if (best != null) {
-                                context.output(best.first());
-                                context.output(best.second());
+                            if (best == null) {
+                                return;
+                            }
+                            final String bestName = best.first().getName();
+                            for (PairedEnds pair : scored) {
+                                opticalDuplicateFinder.addLocationInformation(pair.first().getName(), pair);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: This is kind of strange.  It's modifying the `PairedEnds` but the only place that consumes that information is in `findOpticalDuplicates`  Shouldn't this be called locally in findOpticalDuplicates and not modify this collection?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 7]
```diff
@@ -218,20 +269,48 @@ public void processElement(final ProcessContext context) throws Exception {
...
+                                    record.setAttribute(OPTICAL_DUPLICATE_ATTRIBUTE_NAME, 1);
+                                }
                                 context.output(record);
 
                                 record = pair.second();
-                                record.setIsDuplicate(true);
+                                if (!pairName.equals(bestName)) {
+                                    record.setIsDuplicate(true);
+                                }
+                                if (opticalDuplicateFlags[i]) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: is this duplicated for a reason?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 5 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: It looks like it's not actually important to have the sorted scored list, just to select the best one.  Could this be replaced with:

Optional<PairedEnds> best = paired.get(true).stream().max(Comparator.comparing(PairedEnds::score));

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 7 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: oh, it is duplicated for a reason, scratch that.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 8]
```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
...
+                        public void processElement(final ProcessContext context) throws Exception {
+                            final GATKRead record = context.element();
+                            final SAMFileHeader h = context.sideInput(header);
+                            final String library = LibraryIdGenerator.getLibraryName(h, record.getReadGroup());
+                            DuplicationMetrics metrics = new DuplicationMetrics();
+                            metrics.LIBRARY = library;
+                            // These need to be initialized so that the DuplicationMetrics can be serialized safely.
+                            metrics.PERCENT_DUPLICATION = 0.0;
+                            metrics.ESTIMATED_LIBRARY_SIZE = 0L;
+                            // We have already filtered out secondary and supplementary reads in the partition transform.
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: I'm pretty sure the unmapped reads are filtered out by the partition function as well:

[https://github.com/broadinstitute/hellbender/pull/749/files#diff-05fb5a0326f8c0d57413f307dfbb4187R168]
```
                 public int partitionFor(final GATKRead read, final int n) {
                     return read.isSecondaryAlignment() || read.isSupplementaryAlignment() || read.isUnmapped() ? ReadsPartition.NOT_PRIMARY.ordinal() : ReadsPartition.PRIMARY.ordinal();
                 }
```

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 9]
```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
...
+            public DuplicationMetrics apply(Iterable<DuplicationMetrics> input) {
+                DuplicationMetrics metricsSum = new DuplicationMetrics();
+                int num = 0;
+                for (final DuplicationMetrics metrics : input) {
+                    if (metricsSum.LIBRARY == null) {
+                        metricsSum.LIBRARY = metrics.LIBRARY;
+                    }
+                    // This should never happen, as we grouped by key using library as the key.
+                    if (!metricsSum.LIBRARY.equals(metrics.LIBRARY)) {
+                        return null;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: If this is an error condition it should throw an exception instead I think.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 10]
```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
...
+                    });
+        }
+    }
+
+
+    /**
+     * Special coder for the DuplicationMetrics class.
+     */
+    private static final class DuplicationMetricsCoder extends CustomCoder<DuplicationMetrics> {
+        private static final long serialVersionUID = 1L;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Why this instead of just using SerializableCoder?  I assume you ran into problems with the default serializable coder?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 11]
```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
...
+        private PTransform<PCollection<? extends KV<String, DuplicationMetrics>>, PCollection<KV<String, DuplicationMetrics>>> finalizeMetrics() {
+            return ParDo
+                .named("finalize metrics")
+                .of(new DoFn<KV<String, DuplicationMetrics>, KV<String, DuplicationMetrics>>() {
+                        private static final long serialVersionUID = 1l;
+
+                        @Override
+                        public void processElement(final ProcessContext context) throws Exception {
+                            DuplicationMetrics metrics = context.element().getValue();
+                            metrics.READ_PAIRS_EXAMINED = metrics.READ_PAIRS_EXAMINED / 2;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: could you add a comment about this / 2?  Is it because it's because it's counted for every read when it should be counted for pairs?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 12]
```diff
@@ -344,9 +622,15 @@ private static int score(final GATKRead record) {
     /**
      * Struct-like class to store information about the paired reads for mark duplicates.
      */
-    private static final class PairedEnds {
+    private static final class PairedEnds implements OpticalDuplicateFinder.PhysicalLocation {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Since the values added by `OpticalDuplicateFinder` seem to only be used in OpticalDuplicate finder it would be nice if they didn't have to be exposed here.

File: src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java [hunk 13]
```diff
@@ -51,5 +92,27 @@ public void testMarkDuplicatesDataflowIntegrationTestLocal(final File input, fin
...
+
+        final MetricsFile<DuplicationMetrics, Comparable<?>> metricsOutput = new MetricsFile<>();
+        try {
+            metricsOutput.read(new FileReader(metricsFile));
+        } catch (final FileNotFoundException ex) {
+            System.err.println("Metrics file not found: " + ex);
+        }
+        Assert.assertEquals(metricsOutput.getMetrics().size(), metricsExpected.size(),
+                            "Wrong number of metrics with non-zero fields.");
+        for (int i = 0; i < metricsOutput.getMetrics().size(); i++ ){
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java]: Is this ordering guaranteed or is it based on whatever arbitrary order the `PCollection` of `KV<Library,DuplicationMetric>` happened to be in?

**Comment by lbergelson** [comment]
Since we never actually look to see if something IS an optical duplicate and only care about the total number, we could just output a single annotation on one read in the best read pair with the number of optical duplicates found for that set of reads.  It would make the code simpler but maybe not make as much sense logically?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 14]
```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
...
+        public void processElement(ProcessContext c) throws Exception {
+            File dest = c.element();
+            Map<String,Iterable<DuplicationMetrics>> metrics = c.sideInput(mapView);
+            final MetricsFile<DuplicationMetrics,Double> file = new MetricsFile<>();
+            for (final Map.Entry<String,Iterable<DuplicationMetrics>> entry : metrics.entrySet()) {
+                for (final DuplicationMetrics m : entry.getValue()) {
+                    file.addMetric(m);
+                }
+            }
+            // Should we add a histogram?
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: yup!  and headers!

File: src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java [hunk 15]
```diff
@@ -1,37 +1,78 @@
...
+            // The first two values are total reads and duplicate reads. The list is an encoding of the metrics
+            // file output by this bam file. These metrics files all match the outputs of picard mark duplicates.
+            {new File(MarkDuplicatesIntegrationTest.TEST_DATA_DIR,"example.chr1.1-1K.unmarkedDups.noDups.bam"), 20, 0,
+             ImmutableList.of(ImmutableList.of("Solexa-16419", 0L, 3L, 0L, 0L, 0L, 0L, 0.0, 0L),
+                              ImmutableList.of("Solexa-16416", 0L, 1L, 0L, 0L, 0L, 0L, 0.0, 0L),
+                              ImmutableList.of("Solexa-16404", 0L, 3L, 0L, 0L, 0L, 0L, 0.0, 0L),
+                              ImmutableList.of("Solexa-16406", 0L, 1L, 0L, 0L, 0L, 0L, 0.0, 0L),
+                              ImmutableList.of("Solexa-16412", 0L, 1L, 0L, 0L, 0L, 0L, 0.0, 0L))},
+            {new File(MarkDuplicatesIntegrationTest.TEST_DATA_DIR,"example.chr1.1-1K.unmarkedDups.bam"), 90, 6,
+             ImmutableList.of(ImmutableList.of("Solexa-16419", 4L, 4L, 4L, 0L, 0L, 0L, 0.0, 0L),
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java]: Would it be easier to pass this as a MetricFile?

File: src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java [hunk 16]
```diff
@@ -1,37 +1,78 @@
...
     }
 
     @Test(groups = "dataflow", dataProvider = "md")
-    public void testMarkDuplicatesDataflowIntegrationTestLocal(final File input, final long totalExpected, final long dupsExpected) throws IOException {
+    public void testMarkDuplicatesDataflowIntegrationTestLocal(
+        final File input, final long totalExpected, final long dupsExpected,
+        List<List<String>> metricsExpected) throws IOException {
+
         ArgumentsBuilder args = new ArgumentsBuilder();
-        args.add("--"+StandardArgumentDefinitions.INPUT_LONG_NAME); args.add(input.getPath());
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java]: Why switch from the constant to a hardcoded value?

**Comment by lbergelson** [comment]
@tovanadler Review complete.  Looks good, just a few comments.

I have a few comments about the organization of duplicate marking.  I think you've inherited some very old style code that could maybe use some refactoring.

I think we do need to also include the histogram and the metrics headers.  Those could be done in a separate ticket though.

I'm a bit worried that the test is indeterministic.  Unless I overlooked something which is likely, it seems like it might depend on the ordering of a PCollection which is undefined.  This isn't problematic for the actual metric file, but might be for the tests.

What do you think about reorganizing to output an annotation on only 1 of the "best" reads with the count of all optical duplicates in it's group.  That would simplify the code, and since we only care about the global count it wouldn't change the information content.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java [hunk 17]
```diff
@@ -49,14 +53,39 @@
...
+            "to estimate the rate of optical duplication in order to give a more accurate estimated library size. " +
+            "Set this option to null to disable optical duplicate detection. " +
+            "The regular expression should contain three capture groups for the three variables, in order. " +
+            "It must match the entire read name. " +
+            "Note that if the default regex is specified, a regex match is not actually done, but instead the read name " +
+            " is split on colon character. " +
+            "For 5 element names, the 3rd, 4th and 5th elements are assumed to be tile, x and y values. " +
+            "For 7 element names (CASAVA 1.8), the 5th, 6th, and 7th elements are assumed to be tile, x and y values.",
+            optional = true)
+    public String READ_NAME_REGEX = OpticalDuplicateFinder.DEFAULT_READ_NAME_REGEX;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Could you add a test case with null here?  We're probably going to be changing our handling of null in input at some point and it would be good to know that we'll catch this instance if we do.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 1 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Yup, whoops. Thanks.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 2 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Created issue #776.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 3 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Done here and below. The reason I switched it was that I wanted it to be clear that the keys were different, but I think the comment does that.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 4 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: On that note, I am definitely going to need some help rebasing this thing.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 5 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Good call.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 6 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Agreed that it's strange, that just seems to be the way the class is designed. If I add it to findOpticalDuplicates I have to change all the other uses. I know that class was heavily optimized so I am a bit wary to change it.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 8 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: So is it safe to assume we just don't care too much about those metrics?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 9 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: The issue is that apply doesn't throw an error, so I wasn't sure what else to do.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 10 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Well, I was given errors that it could not be encoded properly when I had no coder specified. Should I have just specified a default one? I don't recall the specific errors, but can check.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 11 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Yup.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 12 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: I know. I tried a whole bunch of ways to reengineer it and this seemed to be the path of least code upheaval. This is how the non-dataflow version does it. Otherwise, there is all sorts of copying going on and I was worried about performance. But if you prefer, I can give that a shot, with a note that we could return to this method if it turns out to be better.

File: src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java (hunk 13 above)
Comment (tovanadler) [inline src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java]: The latter. But wasn't sure how else to test it.

File: src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java (hunk 15 above)
Comment (tovanadler) [inline src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java]: Potentially, but I did like having all the data just listed here and easy to look at when comparing to the metrics file output by the non-dataflow version. Also, wasn't sure metricsfile had the right constructor.

File: src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java (hunk 16 above)
Comment (tovanadler) [inline src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/MarkDuplicatesDataflowIntegrationTest.java]: Artifact of trying to use the SamFileTester as part of this PR. Reverted.

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 14 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: What are the headers? The issue with the histogram is that they do a check of metricsByLibrary.size() = 1, which is pretty hard/ugly to do in dataflow. What is the point of that check?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 9 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: You can always throw a `RuntimeException` like `GatkException` can't you?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 14 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: That's a good question... I'm not sure.  Probably because the histogram only makes sense for a singe library and there isn't a way to represent it for multiple libraries.

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: Althought they could have multiple histograms, one for each library.... I don't know why they don't do that.  Maybe lack of interest?

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: @yfarjoun Can you elucidate?

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: The headers are just a time stamp and the input command line.  But they're formatted in a specific way

```
## htsjdk.samtools.metrics.StringHeader
# org.broadinstitute.hellbender.tools.picard.sam.markduplicates.MarkDuplicates  --INPUT /Users/louisb/Workspace/hellbender/src/test/resources/org/broadinstitute/hellbender/tools/picard/sam/MarkDuplicates/merge1.sam --INPUT /Users/louisb/Workspace/hellbender/src/test/resources/org/broadinstitute/hellbender/tools/picard/sam/MarkDuplicates/merge2.sam --INPUT /Users/louisb/Workspace/hellbender/src/test/resources/org/broadinstitute/hellbender/tools/picard/sam/MarkDuplicates/merge3.sam --OUTPUT /var/folders/xt/vq7wz8955r1401mv8w0f4zf9qbfwzl/T/MarkDuplicates.7414082829862458425.tmp/MarkDuplicates.sam --METRICS_FILE /var/folders/xt/vq7wz8955r1401mv8w0f4zf9qbfwzl/T/MarkDuplicates.7414082829862458425.tmp/MarkDuplicates.duplicate_metrics  --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --reportMemoryStats false --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --REMOVE_DUPLICATES false --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --READ_NAME_REGEX [a-zA-Z0-9]+:[0-9]:([0-9]+):([0-9]+):([0-9]+).* --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 5 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --help false --version false --VERBOSITY INFO --QUIET false
## htsjdk.samtools.metrics.StringHeader
# Started on: August 3, 2015 2:33:08 PM EDT
```

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 17 above)
Comment (tovanadler) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: What is the appropriate behavior?

**Comment by tovanadler** [comment]
Thanks for the review.
Regarding the code complexity, I definitely agree. I was trying to not do too much code surgery. For your suggestion regarding just outputting the total, perhaps if I name it something like total optical duplicates? The issue is that you could have two optical duplicate clusters, but I guess we don't really care, right? We just need the total?
Also, I can rewrite the test to load the information in as a map and then just make sure the sizes are the same and query the map. Would that be better?

File: src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java (hunk 14 above)
Comment (yfarjoun) [inline src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java]: the way the metrics files are _currently_ setup there can only be one histogram (though it can have multiple values) and the headers are fixed in advance. This means that when multiple libraries are present we cannot display the detailed metrics (which are per library)...It would be nice to have a solution that can support multiple histograms in the result. @ktibbett perhaps you would like to chime in as well?

**Comment by tovanadler** [comment]
I just commited some changes based on your comments and also fixed the tests. Can you see them as part of this PR or should I make a separate PR? I am waiting to rebase until we agree on everything in this CL.

**Comment by tovanadler** [comment]
OK, things have become sufficiently complicated here with my revert branches and my difference from master that I am just going to rebase everything now. I will probably have to send you a new PR with the updated changes, if that is ok.

**Comment by lbergelson** [comment]
Rebasing sooner rather than later is a good idea.  Easiest thing to do is to make a copy of your branch and do the rebase in that which seems like what you're planning.  When you're sure it's gone right you can always --reset the old branch to the new branch and just update the existing pull request.
